sam_view.c

```diff
@@ -404,47 +413,243 @@ int parse_aux_list(auxhash_t *h, char *optarg) {
+    reg[i].intervals[reg[i].count - 1].end = end;
+    return reg;
+}
+static void _reglist_merge(hts_reglist_t *reg, int nreg)
+{
+    int i,j;
+    for (i=0; i<nreg; i++)
+    {
+        qsort(reg[i].intervals,reg[i].count,sizeof(*reg[i].intervals),cmp_reglist_intervals);
+        int k = 1;
+        for (j=1; j<reg[i].count; j++)
+        {
+            if ( reg[i].intervals[k-1].end < reg[i].intervals[j].beg )
+            {
+                if ( k < j ) reg[i].intervals[k] = reg[i].intervals[j];
+                k++;
+                continue;
+            }
+            if ( reg[i].intervals[k-1].end < reg[i].intervals[j].end ) reg[i].intervals[k-1].end = reg[i].intervals[j].end;
+        }
+        reg[i].count = k;
+        reg[i].max_end = reg[i].intervals[k-1].end;
+    }
+}
+static int fetch_pairs_collect_mates(samview_settings_t *conf, hts_itr_multi_t *iter)
+{
+    int nunmap = 0, ret = 0, nmates = 0, nreg = iter->n_reg;
+    hts_reglist_t *reglist = _reglist_dup(iter->reg_list, iter->n_reg);
+    void *mate_names = khash_str2int_init();
+
+    bam1_t *rec = bam_init1();
+    while ( sam_itr_multi_next(conf->in, iter, rec)>=0 )
+    {
+        if ( rec->core.mtid>=0 && bed_overlap(conf->bed, sam_hdr_tid2name(conf->header,rec->core.mtid), rec->core.mpos, rec->core.mpos) ) continue;
+
+        nmates++;
+        if ( !khash_str2int_has_key(mate_names,bam_get_qname(rec)) )
+        {
+            ret = khash_str2int_set(mate_names, strdup(bam_get_qname(rec)), 1);
+            if ( ret<0 ) error("[%s:%d] could not expand memory structures, %d elements. (Todo: external sort)\n",__FILE__,__LINE__,nmates);
+        }
+
+        if ( rec->core.mtid < 0 || (rec->core.flag & BAM_FMUNMAP) ) nunmap = 1;
+        if ( rec->core.mtid >= 0 )
+        {
+            reglist = _reglist_push(reglist, &nreg, sam_hdr_tid2name(conf->header,rec->core.mtid), rec->core.mtid, rec->core.mpos,rec->core.mpos+1);
+            if ( !reglist ) error("[%s:%d] could not expand memory structures, %d elements. (Todo: external sort)\n",__FILE__,__LINE__,nmates);
+        }
+    }
+    _reglist_merge(reglist, nreg);
+    if ( nunmap )
+        reglist = _reglist_push(reglist,&nreg,"*",HTS_IDX_NOCOOR,0,HTS_POS_MAX);
+    hts_itr_multi_destroy(iter);
+    iter = sam_itr_regions(conf->hts_idx, conf->header, reglist, nreg);
+    if ( !iter ) error("[%s:%d] iterator could not be created. Aborting.\n",__FILE__,__LINE__);
+    while ( sam_itr_multi_next(conf->in, iter, rec)>=0 )
+    {
+        int drop = 1;
+        if ( rec->core.tid >=0 && bed_overlap(conf->bed, sam_hdr_tid2name(conf->header,rec->core.tid), rec->core.pos, bam_endpos(rec)) ) drop = 0;
+        if ( drop && khash_str2int_get(mate_names,bam_get_qname(rec),&ret)==0 ) drop = 0;
+        if ( !drop && check_sam_write1(conf->out,conf->header,rec,conf->fn_out,&ret) < 0 ) error("aborting");
+    }
+
+    int i;
+    for (i=0; i<nreg; i++) free((char*)reglist[i].reg);
```

Comment:
This can be dropped if we can get rid of all the `strdup()` calls when setting the `reg` fields.